### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by 4ertya**

### CODE REVIEW

File: api/src/main/java/com/epam/esm/controller/CertificatesController.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.List;
+
+@RestController
+@AllArgsConstructor(onConstructor = @__(@Autowired))
+@RequestMapping("/certificates")
+public class CertificatesController {
+
+    private final CertificatesService certificatesService;
+
+    @GetMapping(produces = {"application/json"})
```
Comment (DmitrySay) [inline api/src/main/java/com/epam/esm/controller/CertificatesController.java]: @GetMapping(produces = {"application/json"})  use @GetMapping

File: api/src/main/java/com/epam/esm/controller/CertificatesController.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import lombok.AllArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.*;
+
+import javax.validation.Valid;
+import java.util.List;
+
+@RestController
+@AllArgsConstructor(onConstructor = @__(@Autowired))
```
Comment (DmitrySay) [inline api/src/main/java/com/epam/esm/controller/CertificatesController.java]: Use @RequiredArgsConstructor

File: api/src/main/java/com/epam/esm/controller/CertificatesController.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+@RequestMapping("/certificates")
```
Comment (DmitrySay) [inline api/src/main/java/com/epam/esm/controller/CertificatesController.java]: @RequestMapping("/certificates")  It is nice to add  prefix api or even api/v1 to mapping @RequestMapping("/api/certificates")

File: api/src/main/java/com/epam/esm/controller/CertificatesController.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            @RequestParam(required = false) String tag,
+            @RequestParam(required = false) String name,
+            @RequestParam(required = false) String description,
+            @RequestParam(required = false) String dateSort,
+            @RequestParam(required = false) String nameSort
+    ) {
+        return certificatesService.readAll(tag, name, description, dateSort, nameSort);
+    }
+
+    @GetMapping(value = "/{id}", produces = {"application/json"})
```
Comment (DmitrySay) [inline api/src/main/java/com/epam/esm/controller/CertificatesController.java]: Use @GetMapping "/{id}") . Your way is redundant. Check all controllers.

File: api/src/main/java/com/epam/esm/controller/TagController.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    public List<TagDTO> readAll() {
+        return tagService.readAll();
+    }
+
+    @GetMapping("/{id}")
+    public TagDTO read(@PathVariable("id") int id) {
+        return tagService.read(id);
+    }
+
+    @PostMapping(value = "/new")
```
Comment (DmitrySay) [inline api/src/main/java/com/epam/esm/controller/TagController.java]: @PostMapping(value = "/new") - Pls read RESTful Resource Naming Conventions and correct url

File: core/src/main/java/com/epam/esm/exception/ServiceException.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package com.epam.esm.exception;
+
+public class ServiceException extends RuntimeException {
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/exception/ServiceException.java]: ServiceException ? Looks useless.

File: core/src/main/java/com/epam/esm/mapper/CertificateMapper.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import com.epam.esm.dto.TagDTO;
+import com.epam.esm.model.Certificate;
+import com.epam.esm.model.Tag;
+
+import java.util.List;
+
+public interface CertificateMapper {
+    Certificate toEntity(CertificateDTO certificateDTO);
+
+    CertificateDTO toDto(Certificate certificate, List<TagDTO> tags);
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/mapper/CertificateMapper.java:14]: CertificateDTO toDto(Certificate certificate)

File: core/src/main/java/com/epam/esm/mapper/impl/CertificateMapperImpl.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.List;
+
+
+@Service
+public class CertificateMapperImpl implements CertificateMapper {
+
+    @Override
+    public Certificate toEntity(CertificateDTO certificateDTO) {
+        Certificate certificate = new Certificate(
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/mapper/impl/CertificateMapperImpl.java:21]: certificateDTO ==null - What will happen?

File: core/src/main/java/com/epam/esm/model/Certificate.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+@AllArgsConstructor
+public class Certificate {
+    private Integer id;
+    private String name;
+    private String description;
+    private Integer price;
+    private Integer duration;
+    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS")
+    private LocalDateTime createDate;
+    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS")
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/model/Certificate.java]: @JsonFormat move to DTO

File: core/src/main/java/com/epam/esm/model/Tag.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package com.epam.esm.model;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+public class Tag {
+    Integer id;
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/model/Tag.java]: Long Id

File: core/src/main/java/com/epam/esm/repository/CertificateDAO.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+    List<Certificate> readAll();
+
+    List<Certificate> readAllBySpecification(Specification specification);
+
+    Optional<Certificate> read(int id);
+
+    Optional<Certificate> create(Certificate certificate);
+
+    Optional<Certificate> update(Certificate certificate);
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/CertificateDAO.java]: Optional ?

File: core/src/main/java/com/epam/esm/repository/CertificateDAO.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+    Optional<Certificate> delete(int id);
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/CertificateDAO.java]: Optional?

File: core/src/main/java/com/epam/esm/repository/impl/SQLCertificateTagsDAO.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
+package com.epam.esm.repository.impl;
+
+import com.epam.esm.repository.CertificateTagsDAO;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.stereotype.Repository;
+
+@Repository
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/impl/SQLCertificateTagsDAO.java]: @__(@Autowired)  - correct here

File: core/src/main/java/com/epam/esm/repository/impl/SQLCertificateTagsDAO.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+    private final static String INSERT_QUERY = "INSERT INTO certificates_tags (gift_certificates_id, tags_id) VALUES (?,?) ON CONFLICT(gift_certificates_id, tags_id) DO NOTHING;";
+    private final static String DELETE_QUERY = "DELETE FROM certificates_tags WHERE gift_certificates_id=? AND tags_id=? ";
+    private final static String DELETE_BY_CERTIFICATE_ID_QUERY = "DELETE FROM certificates_tags WHERE gift_certificates_id=?";
+    private final static String DELETE_BY_TAG_ID_QUERY = "DELETE FROM certificates_tags WHERE tags_id=? ";
+
+    private final JdbcTemplate jdbcTemplate;
+
+    @Override
+    public Integer add(int certificateId, int tagId) {
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/impl/SQLCertificateTagsDAO.java]: Integer  1 or 0. Useless.  Use Boolean 1==jdbcTemplate.update  or void

File: core/src/main/java/com/epam/esm/repository/impl/SQLCertificatesDAO.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    @Override
+    public Optional<Certificate> create(Certificate certificate) {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        SqlParameterSource parameterSource = new MapSqlParameterSource()
+                .addValue("name", certificate.getName())
+                .addValue("description", certificate.getDescription())
+                .addValue("price", certificate.getPrice())
+                .addValue("duration", certificate.getDuration());
+        namedParameterJdbcTemplate.update(CREATE_CERTIFICATE, parameterSource, keyHolder, new String[]{"id"});
+        return keyHolder.getKey() != null ? read((Integer) keyHolder.getKey()) : Optional.empty();
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/impl/SQLCertificatesDAO.java]: Optional.empty() ?  and  Optional<Certificate> useless.    return Certificate and set Id !

File: core/src/main/java/com/epam/esm/repository/impl/SQLCertificatesDAO.java [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class SQLCertificatesDAO implements CertificateDAO {
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/impl/SQLCertificatesDAO.java]: SQLCertificatesDAO -  rename CertificatesDAOImpl.  ClassName= InterfaceName+Impl. Check name on all classes.
Or use CertificateRepository and class  CertificateRepositoryImpl

File: core/src/main/java/com/epam/esm/repository/impl/SQLCertificatesDAO.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Override
+    public Optional<Certificate> update(Certificate certificate) {
+        int rows = jdbcTemplate.update(UPDATE_BY_ID_QUERY, certificate.getName(), certificate.getDescription(), certificate.getPrice(), certificate.getDuration(), certificate.getId());
+        return rows == 0 ? Optional.empty() : read(certificate.getId());
+    }
+
+    @Override
+    public Optional<Certificate> delete(int id) {
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/impl/SQLCertificatesDAO.java]: void + jdbcTemplate.update(DELETE_BY_ID_QUERY, id);

File: core/src/main/java/com/epam/esm/repository/impl/SQLTagDAO.java [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Optional;
+
+
+@Repository
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class SQLTagDAO implements TagDAO {
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/impl/SQLTagDAO.java]: rename SQLTagDAO

File: core/src/main/java/com/epam/esm/repository/impl/SQLTagDAO.java [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+                .findAny();
+    }
+
+    @Override
+    public List<Tag> findByCertificateId(int id) {
+        return jdbcTemplate.query(SELECT_BY_CERTIFICATE_ID_QUERY, new Object[]{id}, new BeanPropertyRowMapper<>(Tag.class));
+    }
+
+    @Override
+    public Optional<Tag> create(Tag tag) {
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/impl/SQLTagDAO.java]: Refactoring the same as a certificate

File: core/src/main/java/com/epam/esm/repository/impl/SQLTagDAO.java [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+                .addValue("name", tag.getName());
+        namedParameterJdbcTemplate.update(INSERT_QUERY, parameterSource, keyHolder, new String[]{"id"});
+
+        return keyHolder.getKey() != null ? findById((Integer) keyHolder.getKey()) : Optional.empty();
+    }
+
+    @Override
+    public Optional<Tag> update(Tag tag) {
+        jdbcTemplate.update(UPDATE_QUERY, tag.getName(), tag.getId());
+        return findById(tag.getId());
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/impl/SQLTagDAO.java]: Only jdbcTemplate.update  - return tag or void

File: core/src/main/java/com/epam/esm/repository/impl/SQLTagDAO.java [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Override
+    public Optional<Tag> delete(int id) {
+        Optional<Tag> tag = findById(id);
+        if (tag.isPresent()) {
+            jdbcTemplate.update(DELETE_QUERY, id);
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/impl/SQLTagDAO.java]: same as certificate

File: core/src/main/java/com/epam/esm/repository/specification/SpecificationCreator.java [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+
+import com.epam.esm.dto.DataSortType;
+import com.epam.esm.repository.specification.impl.*;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Service
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/specification/SpecificationCreator.java]: MayBe @Component -?

File: core/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class CertificateServiceImpl implements CertificatesService {
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java]: CertificatesService rename CertificateService

File: core/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+        return certificateMapper.toDto(certificate, tags);
+    }
+
+    @Override
+    @Transactional
+    public CertificateDTO update(int id, CertificateDTO certificateDTO) {
+        Certificate newCertificate = certificateMapper.toEntity(certificateDTO);
+        newCertificate.setId(id);
+        Certificate certificate = certificateDAO.update(newCertificate).orElseThrow(() -> new EntityNotUpdatedException("Certificate", id));
+        if (certificateDTO.getTags() != null) {
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java:92]: Create sub-method

File: core/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+            }
+        }
+        List<TagDTO> tags = tagService.findByCertificateId(id);
+        return certificateMapper.toDto(certificate, tags);
+    }
+
+    @Override
+    @Transactional
+    public CertificateDTO delete(int id) {
+        List<TagDTO> tags = tagService.findByCertificateId(id);
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java]: tagService.findByCertificateId(id); - tagService.findTagsByCertificateId(id) - Try named method to clear understand what method does.

File: core/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+
+    private final TagDAO tagDAO;
+    private final TagMapper tagMapper;
+    private final CertificateTagService certificateTagService;
+
+
+    @Override
+    public List<TagDTO> readAll() {
+        List<Tag> tags = tagDAO.findAll();
+        return tags.isEmpty() ? null : tagMapper.toDtoList(tagDAO.findAll());
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java]: tags.isEmpty() ? null - move  null validation to mapper

File: core/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    public TagDTO create(TagDTO tagDTO) {
+        Tag tag = tagDAO.create(tagMapper.toEntity(tagDTO)).orElseThrow(() -> new EntityNotAddedException("Tag"));
+        return read(tag.getId());
+    }
+
+    @Override
+    public TagDTO update(int id, TagDTO tagDTO) {
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java]: public TagDTO update(int id, TagDTO tagDTO) { }   --->  update (TagDTO tagDTO) set Id to tagDTO before

File: core/src/main/resources/create_db.sql [hunk 28]
```diff
@@ -0,0 +1,38 @@
+create database certificates;
+
+create table if not exists certificates
+(
+    id               serial  not null
+        constraint gift_certificate_pkey
```
Comment (DmitrySay) [inline core/src/main/resources/create_db.sql]: move to the end of columns

File: core/src/main/resources/create_db.sql [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+            primary key,
+    name             text    not null,
+    description      text,
+    price            integer not null         default 0,
```
Comment (DmitrySay) [inline core/src/main/resources/create_db.sql]: default 0 - remove

Comment (DmitrySay) [inline core/src/main/resources/create_db.sql]: price    integer ?        numeric and use BigDecinal for price in model;

**Review by DmitrySay — COMMENTED** [review]
FeedBack

File: core/src/main/java/com/epam/esm/exception/ServiceException.java (hunk 6 above)
Comment (4ertya) [reply core/src/main/java/com/epam/esm/exception/ServiceException.java]: `@ExceptionHandler(ServiceException.class)`
I made one method in AppExceptionHandler without listing all classes.
I can create new Exceptions that extends ServiceException and not touch handler.
Is it bad idea?

File: api/src/main/java/com/epam/esm/controller/CertificatesController.java (hunk 1 above)
Comment (4ertya) [reply api/src/main/java/com/epam/esm/controller/CertificatesController.java]: This is a rudiment from the days of @Controller. =)
Fixed.

File: api/src/main/java/com/epam/esm/controller/CertificatesController.java (hunk 2 above)
Comment (4ertya) [reply api/src/main/java/com/epam/esm/controller/CertificatesController.java]: Done

File: api/src/main/java/com/epam/esm/controller/CertificatesController.java (hunk 3 above)
Comment (4ertya) [reply api/src/main/java/com/epam/esm/controller/CertificatesController.java]: Well, if it's nice...
Done.

File: api/src/main/java/com/epam/esm/controller/CertificatesController.java (hunk 4 above)
Comment (4ertya) [reply api/src/main/java/com/epam/esm/controller/CertificatesController.java]: Checked.

File: api/src/main/java/com/epam/esm/controller/TagController.java (hunk 5 above)
Comment (4ertya) [reply api/src/main/java/com/epam/esm/controller/TagController.java]: Thx. Bad google example

File: core/src/main/java/com/epam/esm/mapper/CertificateMapper.java (hunk 7 above)
Comment (4ertya) [reply core/src/main/java/com/epam/esm/mapper/CertificateMapper.java:14]: Done

File: core/src/main/java/com/epam/esm/mapper/impl/CertificateMapperImpl.java (hunk 8 above)
Comment (4ertya) [reply core/src/main/java/com/epam/esm/mapper/impl/CertificateMapperImpl.java:21]: NPE will happen.
I haven't found a way to bypass the validation yet.
But I will think about it.

File: core/src/main/java/com/epam/esm/model/Certificate.java (hunk 9 above)
Comment (4ertya) [reply core/src/main/java/com/epam/esm/model/Certificate.java]: Copy-Paste from DTO. Sorry)

File: core/src/main/java/com/epam/esm/model/Tag.java (hunk 10 above)
Comment (4ertya) [reply core/src/main/java/com/epam/esm/model/Tag.java]: Yeap

File: core/src/main/java/com/epam/esm/mapper/impl/CertificateMapperImpl.java (hunk 8 above)
Comment (DmitrySay) [reply core/src/main/java/com/epam/esm/mapper/impl/CertificateMapperImpl.java:21]: example

https://mapstruct.org/documentation/stable/reference/html/#basic-mappings

@Override
    public CarDto carToCarDto(Car car) {
        if ( car == null ) {
            return null;
        }

**Pull request merged**